File: src/sdl_cpu_info.erl
%% Copyright (c) 2014-2018, Loïc Hoguin <essen@ninenines.eu>
%%
%% Permission to use, copy, modify, and/or distribute this software for any
%% purpose with or without fee is hereby granted, provided that the above
%% copyright notice and this permission notice appear in all copies.
%%
%% THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
%% WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
%% MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
%% ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
%% WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
%% ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
%% OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

-module(sdl_cpu_info).

-export([get_cpu_cache_line_size/0]).
-export([get_cpu_count/0]).
-export([get_system_ram/0]).
-export([has_3dnow/0]).
-export([has_altivec/0]).
-export([has_avx/0]).
-export([has_avx2/0]).
-export([has_mmx/0]).
-export([has_neon/0]).
-export([has_rdtsc/0]).
-export([has_sse/0]).
-export([has_sse2/0]).
-export([has_sse3/0]).
-export([has_sse41/0]).
-export([has_sse42/0]).

-spec get_cpu_cache_line_size() -> non_neg_integer().
get_cpu_cache_line_size() ->
	esdl2:get_cpu_cache_line_size().

-spec get_cpu_count() -> non_neg_integer().
get_cpu_count() ->
	esdl2:get_cpu_count().

-spec get_system_ram() -> non_neg_integer().
get_system_ram() ->
	esdl2:get_system_ram().

-spec has_3dnow() -> boolean().
has_3dnow() ->
	esdl2:has_3dnow().

-spec has_altivec() -> boolean().
has_altivec() ->
	esdl2:has_altivec().

-spec has_avx() -> boolean().
has_avx() ->
	esdl2:has_avx().

-spec has_avx2() -> boolean().
has_avx2() ->
	esdl2:has_avx2().

-spec has_mmx() -> boolean().
has_mmx() ->
	esdl2:has_mmx().

-spec has_neon() -> boolean().
has_neon() ->
	esdl2:has_neon().

-spec has_rdtsc() -> boolean().
has_rdtsc() ->
	esdl2:has_rdtsc().

-spec has_sse() -> boolean().
has_sse() ->
	esdl2:has_sse().

-spec has_sse2() -> boolean().
has_sse2() ->
	esdl2:has_sse2().

-spec has_sse3() -> boolean().
has_sse3() ->
	esdl2:has_sse3().

-spec has_sse41() -> boolean().
has_sse41() ->
	esdl2:has_sse41().

-spec has_sse42() -> boolean().
has_sse42() ->
	esdl2:has_sse42().

File: c_src/sdl_cpu_info.c
// Copyright (c) 2014-2015, Loïc Hoguin <essen@ninenines.eu>
//
// Permission to use, copy, modify, and/or distribute this software for any
// purpose with or without fee is hereby granted, provided that the above
// copyright notice and this permission notice appear in all copies.
//
// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
// ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
// ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
// OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

#include "esdl2.h"

// get_cpu_cache_line_size

NIF_FUNCTION(get_cpu_cache_line_size)
{
	return enif_make_int(env, SDL_GetCPUCacheLineSize());
}

// get_cpu_count

NIF_FUNCTION(get_cpu_count)
{
	return enif_make_int(env, SDL_GetCPUCount());
}

// get_system_ram

NIF_FUNCTION(get_system_ram)
{
	return enif_make_int(env, SDL_GetSystemRAM());
}

// has_3dnow

NIF_FUNCTION(has_3dnow)
{
	if (SDL_Has3DNow())
		return atom_true;

	return atom_false;
}

// has_avx

NIF_FUNCTION(has_avx)
{
	if (SDL_HasAVX())
		return atom_true;

	return atom_false;
}

// has_altivec

NIF_FUNCTION(has_altivec)
{
	if (SDL_HasAltiVec())
		return atom_true;

	return atom_false;
}

// has_mmx

NIF_FUNCTION(has_mmx)
{
	if (SDL_HasMMX())
		return atom_true;

	return atom_false;
}

// has_rdtsc

NIF_FUNCTION(has_rdtsc)
{
	if (SDL_HasRDTSC())
		return atom_true;

	return atom_false;
}

// has_sse

NIF_FUNCTION(has_sse)
{
	if (SDL_HasSSE())
		return atom_true;

	return atom_false;
}

// has_sse2

NIF_FUNCTION(has_sse2)
{
	if (SDL_HasSSE2())
		return atom_true;

	return atom_false;
}

// has_sse3

NIF_FUNCTION(has_sse3)
{
	if (SDL_HasSSE3())
		return atom_true;

	return atom_false;
}

// has_sse41

NIF_FUNCTION(has_sse41)
{
	if (SDL_HasSSE41())
		return atom_true;

	return atom_false;
}

// has_sse42

NIF_FUNCTION(has_sse42)
{
	if (SDL_HasSSE42())
		return atom_true;

	return atom_false;
}
